src/lexer/star_expand.c

```diff
@@ -0,0 +1,129 @@
+/* ************************************************************************** */
+/*                                                                            */
+/*                                                        :::      ::::::::   */
+/*   star_expand.c                                      :+:      :+:    :+:   */
+/*                                                    +:+ +:+         +:+     */
+/*   By: lporoshi <lporoshi@student.42.fr>          +#+  +:+       +#+        */
+/*                                                +#+#+#+#+#+   +#+           */
+/*   Created: 2023/12/15 12:55:50 by lporoshi          #+#    #+#             */
+/*   Updated: 2023/12/18 15:34:22 by lporoshi         ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "libft.h"
+#include "lexer.h"
+
+static char	**get_files_in_cur_dir(void)
+{
+	DIR				*d;
+	struct dirent	*dir;
+	char			**files;
+	int				files_count;
+
+	dir = NULL;
+	files_count = count_files_in_cur_dir();
+	d = opendir(".");
+	if (!d)
+		return (NULL);
+	files = (char **)ft_calloc(files_count + 1, sizeof(char *));
+	if (files == NULL)
+		return (NULL);
+	scan_files_in_dir(&files, files_count, d, dir);
+	closedir(d);
+	return (files);
+}
+
+static char	**compose_expansion_arr(char *pattern, char **candidates)
+{
+	char	**expansion;
+	char	**expansion_orig;
+	int		i;
+	int		sz;
+
+	i = 0;
+	sz = 0;
+	while (candidates[i] != NULL)
+		if (match_wildcard(candidates[i++], pattern) == FT_TRUE)
+			sz++;
+	expansion = (char **)ft_calloc(sz + 2, sizeof(char *));
+	if (expansion == NULL)
+		return (NULL);
+	expansion_orig = expansion;
+	i = 0;
+	while (candidates[i] != NULL)
+	{
+		if (match_wildcard(candidates[i], pattern) == FT_TRUE)
+			*(expansion++) = ft_strdup(candidates[i]);
+		i++;
+	}
+	return (expansion_orig);
+}
+
+static char	**expand_star_string(char *pattern)
+{
+	char	**files;
+	char	**expansion;
+
+	files = get_files_in_cur_dir();
+	if (files == NULL)
+		return (NULL);
+	expansion = compose_expansion_arr(pattern, files);
+	free_str_arr(&files);
+	return (expansion);
+}
+
+static t_list	*expand_star_token(t_token *tok)
+{
+	t_list	*expanded_lst;
+	t_list	*expanded_entry;
+	char	**expansion_list;
+	int		i;
+
+	expanded_lst = NULL;
+	expansion_list = expand_star_string(tok->token_string);
+	if (expansion_list == NULL)
+		return (NULL);
+	i = 0;
+	while (expansion_list[i] != NULL)
+	{
+		expanded_entry = expanded_star_to_token(expansion_list[i++]);
+		if (expanded_entry == NULL)
+		{
+			ft_lstclear(&expanded_lst, del_token);
+			free_str_arr(&expansion_list);
+			return (NULL);
+		}
+		ft_lstadd_back(&expanded_lst, expanded_entry);
+	}
+	free_str_arr(&expansion_list);
+	return (expanded_lst);
+}
+
+int	expand_all_stars(t_list **tok_lst)
```

Comment:
cd folder && echo *
also I think start expansion should be done in parser not in tokenizer

Reply:
How exactly will it work?
There's a function expand_star_string():
https://github.com/taaae/minishell/blob/bccb5d173360be1fd84bf7054501c446339489db/src/lexer/star_expand.c#L62
that takes an expansion pattern and returns null-terminated array of strings with all the filenames (in current directory) that match that pattern. WIll it suffice? If so, I can just remove other unnecessary functions.
But if you need any other fucntionality, then please attach a prototype/documentation of the function that you need, and I'll write the implementation of it.

Reply:
I'd just forget about wildcards while tokenization and treat it as a normal character (part of TOK_WORD). And expand_star_string signature is good enough imo

Reply:
Okay. Removed this from lexer module.